File: Day02.hs
import Data.List
import Data.Char
import Debug.Trace

test = readFile "test02.txt" >>= print . two
main = readFile "input02.txt" >>= print . two

split f s =
    case break f s of
        ([], _:b) -> split f b
        ( a,  []) -> [a]
        ( a, _:b) -> [a] ++ split f b

data Color = Red | Blue | Green deriving (Show, Eq)

parse "red" = Red
parse "blue" = Blue
parse "green" = Green

type Game = (Int, [Round])

type Round = [(Int, Color)]

pair :: String -> (Int, Color)
pair (' ':s) = (read n, parse color)
  where (n, ' ':color) = break (== ' ') s

line :: String -> [Round]
line = map (map pair . split (== ',')) . split (== ';') . tail . dropWhile (/= ':') 

one = sum . map fst . filter (wouldBePossibleWith 12 13 14) . zip[1..] . map line . lines

wouldBePossibleWith :: Int -> Int -> Int -> Game -> Bool
wouldBePossibleWith r g b (_, rounds) =
    all (valid r g b) rounds

valid r g b round = all (\(n, c) -> 
        n <= (case c of
            Red -> r
            Green -> g
            Blue -> b)
    ) round

two = sum . map power . zip[1..] . map line . lines

power (_, rounds) =
    (f Red rounds * f Green rounds * f Blue rounds)

f :: Color -> [Round] -> Int
f col =
    maximum . concatMap (map (\(n, c) -> if c == col then n else 0))
    

File: zig/day02.zig
const std = @import("std");

pub fn main() !void {
    std.debug.print("total: {d}\n", .{solve(@embedFile("input02.txt"))});
}

fn solve(input: []const u8) u32 {
    var lines = std.mem.split(u8, input, "\n");
    var total: u32 = 0;

    while (lines.next()) |line| {
        var red: u32 = 0;
        var green: u32 = 0;
        var blue: u32 = 0;
        if (line.len == 0) {
            break;
        }
        var y = std.mem.split(u8, line, ":");
        _ = y.next() orelse return 0;
        var round = y.next() orelse return 0;
        var games = std.mem.split(u8, round, ";");
        while (games.next()) |game| {
            var moves = std.mem.split(u8, game, ",");
            while (moves.next()) |move| {
                var parts = std.mem.split(u8, move[1..], " ");
                var ns = parts.next() orelse return 0;
                var n = std.fmt.parseInt(u32, ns, 10) catch return 0;
                var c = parts.next() orelse return 0;
                if (std.mem.eql(u8, c, "red")) {
                    red = @max(red, n);
                }
                if (std.mem.eql(u8, c, "green")) {
                    green = @max(green, n);
                }
                if (std.mem.eql(u8, c, "blue")) {
                    blue = @max(blue, n);
                }
            }
        }
        std.debug.print("r:{d} g:{d} b:{d}\n", .{red, green, blue});
        total += red * blue * green;
        std.debug.print("total: {d}\n", .{total});
    }

    return total;
}
